File: sdk/@launchdarkly/observability-shared/src/sampling/sampleLogs.ts
import { ReadableLogRecord } from '@opentelemetry/sdk-logs'
import { Attributes } from '@opentelemetry/api'
import { merge } from '@opentelemetry/core'
import { ExportSampler } from './ExportSampler'

export function cloneLogRecordWithAttributes(
	log: ReadableLogRecord,
	attributes: Attributes,
): ReadableLogRecord {
	const cloned = {
		...log,
		attributes: merge(log.attributes, attributes),
	}
	return cloned
}

export function sampleLogs(
	items: ReadableLogRecord[],
	sampler: ExportSampler,
): ReadableLogRecord[] {
	if (!sampler.isSamplingEnabled()) {
		return items
	}

	const sampledLogs: ReadableLogRecord[] = []

	for (const item of items) {
		const sampleResult = sampler.sampleLog(item)
		if (sampleResult.sample) {
			if (sampleResult.attributes) {
				sampledLogs.push(
					cloneLogRecordWithAttributes(item, sampleResult.attributes),
				)
			} else {
				sampledLogs.push(item)
			}
		}
		// If not sampled, we simply don't include it in the result
	}

	return sampledLogs
}

File: sdk/@launchdarkly/observability-android/lib/src/main/kotlin/com/launchdarkly/observability/sampling/SampleLogs.kt
package com.launchdarkly.observability.sampling

import io.opentelemetry.api.common.Attributes
import io.opentelemetry.api.common.Value
import io.opentelemetry.sdk.logs.data.LogRecordData

fun sampleLogs(
    items: List<LogRecordData>,
    sampler: ExportSampler
): List<LogRecordData> {

    if (!sampler.isSamplingEnabled()) {
        return items
    }

    val sampledLogs = mutableListOf<LogRecordData>()

    for (item in items) {
        val sampleResult = sampler.sampleLog(item)
        if (sampleResult.sample) {
            if (sampleResult.attributes != null) {
                sampledLogs.add(
                    cloneLogRecordWithAttributes(item, sampleResult.attributes)
                )
            } else {
                sampledLogs.add(item)
            }
        }
        // If not sampled, we simply don't include it in the result
    }

    return sampledLogs
}

private fun cloneLogRecordWithAttributes(
    log: LogRecordData,
    attributes: Attributes
): LogRecordData {
    return object : LogRecordData by log {
        override fun getBodyValue(): Value<*>? = log.getBodyValue()
        override fun getEventName(): String? = log.eventName

        override fun getAttributes(): Attributes {
            val builder = Attributes.builder()
            builder.putAll(log.attributes)
            builder.putAll(attributes)
            return builder.build()
        }
    }
}